pipelines/common/src/main/java/org/openmrs/analytics/JdbcConnectionPools.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openmrs.analytics;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import com.mchange.v2.c3p0.ComboPooledDataSource;
+import java.beans.PropertyVetoException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.sql.DataSource;
+import org.openmrs.analytics.model.DatabaseConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JdbcConnectionPools {
+  private static final Logger log = LoggerFactory.getLogger(JdbcConnectionPools.class);
+
+  private static final ConcurrentMap<DataSourceConfig, DataSource> dataSources =
+      new ConcurrentHashMap<>();
+
+  // This class should not be instantiated!
+  private JdbcConnectionPools() {}
+
+  /**
+   * Creates a new connection pool for the given config or return an already created one if one
+   * exists. The pool size parameters are just hints and won't be used if a pool already exists for
+   * the given configuration. This method is to impose a Singleton pattern for each config.
+   *
+   * @param config the JDBC connection information
+   * @param initialPoolSize initial pool size if a new pool is created.
+   * @param jdbcMaxPoolSize maximum pool size if a new pool is created.
+   * @return
+   */
+  public static synchronized DataSource getPooledDataSource(
+      DataSourceConfig config, int initialPoolSize, int jdbcMaxPoolSize)
+      throws PropertyVetoException {
+    if (!dataSources.containsKey(config)) {
+      dataSources.put(config, createNewPool(config, initialPoolSize, jdbcMaxPoolSize));
+    }
+    return dataSources.get(config);
+  }
```

Comment:
Do we need `ConcurrentHashMap` when we have `static synchronized`?  Also, `static synchronized` will put a lock on complete class disabling accessing anything other than `DataSource` like `DataSourceConfig `. Although, if we merge those two methods into one then we should be good.

Reply:
Right `ConcurrentHashMap` is not really needed; I replaced it with a normal map. I think what  I had in mind was to use `ConcurrentMap.computeIfAbsent` and not use synchronization but given that all access to `dataSources` is `synchronized` we don't really need `ConcurrentMap` (BTW, I prefer this pattern over the double-check+`volatile` approach suggested in the link you mentioned; synchronization issues can quickly get complicated and I prefer to keep it simple/clear).

Re. your other comment, as you mentioned, performance is not that important in this class (it is called once per worker) but I am not sure if I get your point re. _"put a lock on complete class disabling accessing anything other than `DataSource` like `DataSourceConfig`"_? So the lock would be only on this method (and anything else `synchronized`). So for example `dbConfigToDataSourceConfig()` calls are not blocked, right? Or did I misunderstood your point?

Reply:
Thanks for the clarification!